### PR TITLE
Add gradient accumulation rewriter

### LINKED ISSUE #6750: [SPMD] SPMD supports Gradient Accumulation in data parallel
When using SPMD for data parallelism and applying gradient accumulation, the current version of XLA can introduce additional all-reduce operations. Is there a plan to support gradient accumulation under SPMD, similar to what is discussed in this GitHub issue: https://github.com/alpa-projects/alpa/issues/72? @tensorflower-gardener

### PR DESCRIPTION
**Pull request by baoleai**
This commit removes unnecessary allreduce from the data parallel backward process when gradient accumulation is turned on (resolve #6750 ).

When implementing data parallelism using SPMD, SPMD inserts allreduce communication in the backward process, regardless of whether gradient accumulation is turned on or not. However, when using gradient accumulation, only global synchronization of gradients (via allreduce between DP ranks) is needed in iterations where the optimizer updates the weights. In non-updated weight iterations, only local gradient accumulation is needed without additional allreduce to synchronize the gradients.

To recognize backward processes in HLO, users are required to set the op_name containing "backward" to the instruction's metadata. For example, `XLA_IR_DEBUG=1 XLA_HLO_DEBUG=1` needs to be set in torch_xla.

### CODE REVIEW

**Comment by baoleai** [comment]
Hi @bchetioui , I have submitted an implementation. Could you please help review it?

**Review by cheshire — CHANGES_REQUESTED** [review]
This seems to heavily rely on framework state invisible to XLA, shouldn't such changes be in PT/XLA bridge instead?

**Comment by bchetioui** [comment]
Thanks for the PR! Now I think I understand better what you meant by multiple submodules: it seems that your example produces one HLO module per loop iteration, as opposed to a single HLO module for the whole loop---is that right?

At a glance, I'm a little skeptical about accepting such a change on principle. The rewrite you propose does not seem semantics-preserving (unless I missed something), and thus does not seem correct for a single HLO module. Would it be possible to rewrite the Python code such that the whole loop is compiled in one single module? In that case, I'd expect that the existing infrastructure (with `WhileLoopAllReduceCodeMotion`) should already take care of removing unnecessary `all-reduce`s. WDYT?

Another solution could also perhaps be to add manual sharding annotations to your loop body (something akin to JAX's `shard_map`. I am not familiar with TorchXLA.)? Adding @frgossen to the discussion, who should have a better understanding of TorchXLA than me.

**Comment by baoleai** [comment]
Thanks for your suggestions @bchetioui. I also agree that the optimal solution is to enable torch_xla to generate a while loop within a single HLO module, allowing us to utilize WhileLoopAllReduceCodeMotion. Further investigation is needed to understand why the current version is unable to generate a while loop and instead produces multiple HLOs. Moreover, I am uncertain whether enabling GA(gradient accumulation) in TensorFlow and Jax would result in a single HLO module for the while loop.

**Comment by bchetioui** [comment]
@baoleai In JAX, you would produce a single HLO module containing a while loop by using [https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html](jax.lax.while_loop). By using the native Python while loop under a call to `jax.jit`, you would also get a single module, but your loop would be fully unrolled.

Fundamentally, the only way you do not get a single module for the whole while loop is if your `jit` is under the while loop.

**Comment by alanwaketan** [comment]
> Thanks for your suggestions @bchetioui. I also agree that the optimal solution is to enable torch_xla to generate a while loop within a single HLO module, allowing us to utilize WhileLoopAllReduceCodeMotion. Further investigation is needed to understand why the current version is unable to generate a while loop and instead produces multiple HLOs. Moreover, I am uncertain whether enabling GA(gradient accumulation) in TensorFlow or Jax would result in a single HLO module for the while loop.

We have ongoing work w.r.t this. Feel free to ping me offline.

**Comment by frgossen** [comment]
> Further investigation is needed to understand why the current version is unable to generate a while loop and instead produces multiple HLOs.

Most likely some graph break. There is little XLA can do when it gets too small HLO modules. I believe the right place to fix this is in the bridge.
